### ISSUE TITLE
Authenticator/Spawner configuration cooperation

### ISSUE THREAD

**Issue by vilhelmen** [jupyterhub/jupyterhub#2261]
I'm looking at developing some tighter integration between the authenticator and the spawner and wondering if it fits within the scope of the main project contribution-wise. I'd hate to hide useful features away in custom forks.

What I'm proposing is a designated section of `auth_state` for storing user profile data relevant to the more complex spawners that aids in user configuration for things like mounted homedir disk permissions.

In particular, I'm looking at the interaction between PAMAuthenticator/LDAPAuthenticator, and the Kubernetes (or Docker) spawner. At the moment, PAM/LDAP authenticator has easy access to information such as UID, primary GID, and secondary groups that at the moment has to be looked up all over again even though the authenticator most likely already did and discarded the results.

The current Docker/Kubernetes documentation seems to suggest that all files for all users be set to 1000:1000 and rely on container mount isolation, but this isn't feasible when users need to work on a shared machine with mounted homedirs.

This could be achieved by, for example, `Authenticator.build_profile` (`build_spawn_profile`?) which, like `is_admin` #2244, is called during `get_authenticated_user` and returns this designated `auth_state` subdict and merges it with any `auth_state` generated by `authenticate`. From there, any spawner updated to be made aware of this data can use it for spawning. This would cost likely need to be gated by config options like `c.Authenticator.generate_spawn_profile` and `c.Spawner.use_spawn_profile` or something.

I was thinking something along the lines of UID, primary GID, and a GID map laid out like so:
```
auth_state = {
  'spawn_profile': {
    'uid': 1001,
    'primary_gid': [2000, 'jh_user']
    'groups': {
      2000: 'jh_user',
      3000: 'jh_admin',
      4000: 'postgres_user',
      5000: 'stats_3050'
    },
}
```

Then Docker/kubernetes spawners and containers can use this information for things like prepopulating the container with the user's groups and things like `c.KubeSpawner.uid` or `c.KubeSpawner.fs_gid`

**Comment by minrk** [comment]
The interactions between Spawners and Authenticators is tricky, since the answer can be different for every combination of Spawner and Authenticator. There haven't been too many generic cases found that aren't tied pretty specifically, not just to a specific Spawner + Authenticator, but Spawner + Authenticator + individual deployment. I'm always happy to explore possible generic use cases that can be promoted up to the base implementation, though! For instance, it may be that we should promote these uid, gid to traits on the base Spawner class, and encourage all Spawners to support them. We've done this before with Spawner-developed options like `hub_connect_ip`.

With containers, it and can even be sensitive to the *image*. e.g. it is common for docker images to only work properly if the uid/gid are the values already set by the image (this is the default behavior for DockerSpawner). It is not a generally valid assumption that the Authenticator knows what uid or gid(s) the Spawner should use. Only if the container is mounting a volume that *also* uses PAM/LDAP uids/gids would this be right, and the Authenticator doesn't generally have that kind of information, but the deployment admin certainly does. While the concepts of uid and gids are technically generic, the right answer tends to be more sensitive to the spawner choice than the authenticator choice.

I think it would be really cool to prototype something like this in a specific `pre_spawn_start` hook for your deployment, which should be able to do everything described here since it's just mapping auth_state fields to Spawner attributes, and then eventually see what makes sense to be promoted to default behavior or not.

**Comment by vilhelmen** [comment]
I completely agree. It is certainly hard to keep things generic with the kind of integration I'm talking about. Hopefully we can at least expand the use of `auth_state` and suggest at integrations such as an addition to the container startup scripts that can create and configure groups inside the container.

I haven't used `pre_spawn_hook` for much more than a homedir check, but it sounds like a very good place to put an `auth_state` consumer in this case.

I'll move some of my files around and put together a WIP/POC PR to track things. Right now I'm focusing on an LDAP authenticator since our org seems to be moving in a direction that makes PAM difficult (though I'm considering the [dark arts of AD via PAM in a container](https://jhrozek.wordpress.com/2015/03/31/authenticating-a-docker-container-against-hosts-unix-accounts/)). The way the current LDAPAuthenticator works comes off as really strange to me and I'm hesitant to spend time modifying it or creating a new one that has to keep backwards compatibility. A problem for later.

**Comment by vilhelmen** [comment]
@minrk Good news! I got a v1 up and running Friday evening. Feels good to see it all work out after working on it **every day** for the last few weeks. Our sudden change from docker to k8s didn't help. I've thrown everything together over at https://github.com/vilhelmen/jh-k8-etc since it's a handful of changes across repositories. I documented the changes and their impact and left a few TODOs in the readme on the repo so I don't write you another essay.

Take a look at the ldap code over [here](https://github.com/vilhelmen/ldapauthenticator/tree/profile_rewrite), the one linked in the repo is a smashed together version to work in stock 0.9.4 and shouldn't really be used since I broke black/whitelisting to get it done quickly. This shows what I was thinking for the `spawn_profile` section of `auth_state`. `build_profile` would just be another method like `is_admin` to supplement authentication flow without having to jam everything in `authenticate` as we do now.

In the end, we end up with a singlueuser server running with proper uid/gids and select user groups pulled from ldap and serving a user's homedir without smashing up disk permissions!

I really think `build_profile` would make a good addition to the `get_authenticated_user` workflow. The LDAP Authenticator I made is a good lookup-and-bind authenticator (though I may be biased 😇). It could use some additional flexibility tweaks before going live. I think my `GROUP_BUILD` script and configuration could be added to the base image with some minor tweaks and could be useful in general. My changes to the start script are weird, but it has to be done that way for rootsquash. Perhaps `ROOTSQUASH_ENABLED` could be a flag to switch behavior in the weird spots. Unfortunately, there's no real way to execute hooks in the user's homedir with rootsquash without calling a user start script that switches directories and runs another hook directory before calling `$cmd`.

Let me know what you think. I'm pretty happy with it overall.

Is there an ETA for 1.0.0? Our big architecture change/k8s cutover is going to be over winter break and I'm hoping to get `is_admin`, #2268 (which I need to review and fix this week), and ideally `build_profile` in and released by then, but I've got it working well enough for 0.9.4 if not.

**Comment by minrk** [comment]
The main new thing for 1.0 is oauth scopes, which I haven't been able to find time to getting underway, since it is a pretty big task. I may decide to hold that for 2.0, and release 1.0 soon with what we have.

**Comment by vilhelmen** [comment]
I really only know oauth in theory, so I can't help too much there, unfortunately.

I've had an idea, though. What if we pivoted the `build_profile` function to a more generic `post_auth_hook` system? Call them right before the return in `get_authenticated_user` and pass the full `authenticated` dict? Or maybe just `auth_state` to protect against username/admin status changes?

Keeps it pretty generic instead of the more k8s/docker spawner-focused direction I was heading in. Sounds like a good move to me!

**Comment by vilhelmen** [comment]
I've updated https://github.com/vilhelmen/jh-k8-etc with the details of our production cluster. It's a hastily documented mess copied off our internal servers, but it's a good rough example of how to get some complex mods done to user containers.

**Comment by manics** [comment]
I'm closing this as it's quite old and there's no concrete action point. If you're still having problems please post on the [Jupyter Community Forum](https://discourse.jupyter.org/) or open a new issue. Thanks!
